File: stock_sync/stock_sync/report/external_stock_view/external_stock_view.py
# Copyright (c) 2025, Pal Shah and contributors
# For license information, please see license.txt

# stock_sync/report/external_stock_view/external_stock_view.py

import frappe
from frappe import _
from frappe.utils import flt, cint, getdate, nowdate

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    
    # Add summary row
    if data:
        summary = get_summary(data)
        data.append([])  # Empty row
        data.append(summary)
    
    return columns, data

def get_columns():
    return [
        {
            "fieldname": "item_code",
            "label": _("Item Code"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "item_name",
            "label": _("Item Name"),
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "source_site",
            "label": _("Source Site"),
            "fieldtype": "Link",
            "options": "Site Connection",
            "width": 120
        },
        {
            "fieldname": "warehouse",
            "label": _("Warehouse"),
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "actual_qty",
            "label": _("Actual Qty"),
            "fieldtype": "Float",
            "width": 90,
            "precision": 2
        },
        {
            "fieldname": "reserved_qty",
            "label": _("Reserved Qty"),
            "fieldtype": "Float",
            "width": 90,
            "precision": 2
        },
        {
            "fieldname": "ordered_qty",
            "label": _("Ordered Qty"),
            "fieldtype": "Float",
            "width": 90,
            "precision": 2
        },
        {
            "fieldname": "available_qty",
            "label": _("Available Qty"),
            "fieldtype": "Float",
            "width": 90,
            "precision": 2
        },
        {
            "fieldname": "last_sync",
            "label": _("Last Sync"),
            "fieldtype": "Datetime",
            "width": 140
        },
        {
            "fieldname": "age",
            "label": _("Age (hours)"),
            "fieldtype": "Float",
            "width": 80,
            "precision": 1
        }
    ]

def get_data(filters):
    conditions = get_conditions(filters)
    
    query = """
        SELECT
            esv.item_code,
            esv.item_name,
            esv.source_site,
            esv.warehouse,
            esv.actual_qty,
            esv.reserved_qty,
            esv.ordered_qty,
            esv.available_qty,
            esv.last_sync,
            TIMESTAMPDIFF(HOUR, esv.last_sync, NOW()) as age
        FROM `tabExternal Stock View` esv
        WHERE esv.docstatus = 0
        {conditions}
        ORDER BY esv.source_site, esv.item_code, esv.warehouse
    """.format(conditions=conditions)
    
    data = frappe.db.sql(query, filters, as_dict=1)
    
    return data

def get_conditions(filters):
    conditions = []
    
    if filters.get("source_site"):
        conditions.append("esv.source_site = %(source_site)s")
    
    if filters.get("item_code"):
        conditions.append("esv.item_code = %(item_code)s")
    
    if filters.get("warehouse"):
        conditions.append("esv.warehouse = %(warehouse)s")
    
    if filters.get("last_sync_from"):
        conditions.append("esv.last_sync >= %(last_sync_from)s")
    
    if filters.get("last_sync_to"):
        conditions.append("esv.last_sync <= DATE_ADD(%(last_sync_to)s, INTERVAL 1 DAY)")
    
    if filters.get("show_only_available"):
        conditions.append("esv.available_qty > 0")
    
    return " AND " + " AND ".join(conditions) if conditions else ""

def get_summary(data):
    total_items = len(data)
    total_actual_qty = sum([flt(d.actual_qty) for d in data])
    total_reserved_qty = sum([flt(d.reserved_qty) for d in data])
    total_ordered_qty = sum([flt(d.ordered_qty) for d in data])
    total_available_qty = sum([flt(d.available_qty) for d in data])
    
    # Count sites
    sites = set([d.source_site for d in data])
    
    return {
        "item_code": f"<b>{_('SUMMARY')}</b>",
        "item_name": f"<b>{total_items} {_('items from')} {len(sites)} {_('sites')}</b>",
        "actual_qty": f"<b>{total_actual_qty:,.2f}</b>",
        "reserved_qty": f"<b>{total_reserved_qty:,.2f}</b>",
        "ordered_qty": f"<b>{total_ordered_qty:,.2f}</b>",
        "available_qty": f"<b>{total_available_qty:,.2f}</b>",
        "indent": 0,
        "bold": 1,
        "background_color": "#f0f0f0"
    }

File: stock_sync/stock_sync/report/external_stock_view/external_stock_view.js
// Copyright (c) 2025, Pal Shah and contributors
// For license information, please see license.txt

// frappe.query_reports["External Stock View"] = {
// 	"filters": [

// 	]
// };

// stock_sync/report/external_stock_view/external_stock_view.js

frappe.query_reports["External Stock View"] = {
    "filters": [
        {
            "fieldname": "source_site",
            "label": __("Source Site"),
            "fieldtype": "Link",
            "options": "Site Connection",
            "width": "100"
        },
        {
            "fieldname": "item_code",
            "label": __("Item Code"),
            "fieldtype": "Data",
            "width": "100"
        },
        {
            "fieldname": "warehouse",
            "label": __("Warehouse"),
            "fieldtype": "Data",
            "width": "100"
        },
        {
            "fieldname": "last_sync_from",
            "label": __("Last Sync From"),
            "fieldtype": "Date",
            "width": "80",
            "default": frappe.datetime.add_days(frappe.datetime.get_today(), -7)
        },
        {
            "fieldname": "last_sync_to",
            "label": __("Last Sync To"),
            "fieldtype": "Date",
            "width": "80",
            "default": frappe.datetime.get_today()
        },
        {
            "fieldname": "show_only_available",
            "label": __("Show Only Available Stock"),
            "fieldtype": "Check",
            "default": 0
        }
    ],
    
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        // Color code based on available quantity
        if (column.fieldname == "available_qty") {
            if (data.available_qty <= 0) {
                value = `<span style="color: red; font-weight: bold;">${value}</span>`;
            } else if (data.available_qty < 5) {
                value = `<span style="color: orange; font-weight: bold;">${value}</span>`;
            } else {
                value = `<span style="color: green; font-weight: bold;">${value}</span>`;
            }
        }
        
        // Color code for last sync time
        if (column.fieldname == "last_sync") {
            var sync_date = frappe.datetime.str_to_obj(data.last_sync);
            var now = frappe.datetime.str_to_obj(frappe.datetime.now_datetime());
            var diff_hours = frappe.datetime.get_hour_diff(now, sync_date);
            
            if (diff_hours > 24) {
                value = `<span style="color: red; font-weight: bold;">${value}</span>`;
            } else if (diff_hours > 12) {
                value = `<span style="color: orange; font-weight: bold;">${value}</span>`;
            } else {
                value = `<span style="color: green; font-weight: bold;">${value}</span>`;
            }
        }
        
        // Make item code clickable
        if (column.fieldname == "item_code" && data.item_code) {
            value = `<a href="/app/item/${data.item_code}" style="text-decoration: none;">${value}</a>`;
        }
        
        return value;
    },
    
    "onload": function(report) {
        // Add custom button to refresh data from sites
        report.page.add_inner_button(__("Refresh All Sites"), function() {
            frappe.call({
                method: "stock_sync.api.fetch_all_sites",
                args: {},
                callback: function(r) {
                    if (r.message && r.message.success) {
                        frappe.show_alert({
                            message: __("Refreshed {0} sites. {1} successful, {2} failed", 
                                [r.message.summary.total_sites, 
                                 r.message.summary.successful, 
                                 r.message.summary.failed]),
                            indicator: "green"
                        });
                        report.refresh();
                    } else {
                        frappe.msgprint({
                            title: __("Refresh Failed"),
                            message: r.message.error || __("Unknown error"),
                            indicator: "red"
                        });
                    }
                },
                error: function(e) {
                    frappe.msgprint({
                        title: __("Error"),
                        message: e.message,
                        indicator: "red"
                    });
                }
            });
        }, __("Actions"));
        
        // Add custom button to refresh specific site
        report.page.add_inner_button(__("Refresh Selected Site"), function() {
            var filters = report.get_values();
            if (filters.source_site) {
                frappe.call({
                    method: "stock_sync.api.fetch_from_site",
                    args: {
                        site_name: filters.source_site
                    },
                    callback: function(r) {
                        if (r.message && r.message.success) {
                            frappe.show_alert({
                                message: __("Refreshed {0} items from {1}", 
                                    [r.message.count, filters.source_site]),
                                indicator: "green"
                            });
                            report.refresh();
                        } else {
                            frappe.msgprint({
                                title: __("Refresh Failed"),
                                message: r.message.error || __("Unknown error"),
                                indicator: "red"
                            });
                        }
                    }
                });
            } else {
                frappe.msgprint({
                    title: __("Warning"),
                    message: __("Please select a Source Site first"),
                    indicator: "orange"
                });
            }
        }, __("Actions"));
    }
};